types/math/dec_test.go

```diff
@@ -89,28 +137,437 @@ func TestDec(t *testing.T) {
+
+	b, err := a.Add(zero)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(b))
+}
+
+// Property: a + b == b + a
+func testAddCommutative(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	b := genDec.Draw(t, "b").(Dec)
+
+	c, err := a.Add(b)
+	require.NoError(t, err)
+
+	d, err := b.Add(a)
+	require.NoError(t, err)
+
+	require.True(t, c.IsEqual(d))
+}
+
+// Property: (a + b) + c == a + (b + c)
+func testAddAssociative(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	b := genDec.Draw(t, "b").(Dec)
+	c := genDec.Draw(t, "c").(Dec)
+
+	// (a + b) + c
+	d, err := a.Add(b)
+	require.NoError(t, err)
+
+	e, err := d.Add(c)
+	require.NoError(t, err)
+
+	// a + (b + c)
+	f, err := b.Add(c)
+	require.NoError(t, err)
+
+	g, err := a.Add(f)
+	require.NoError(t, err)
+
+	require.True(t, e.IsEqual(g))
+}
+
+// Property: a - 0 == a
+func testSubRightIdentity(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	zero := NewDecFromInt64(0)
+
+	b, err := a.Sub(zero)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(b))
+}
+
+// Property: a - a == 0
+func testSubZero(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	zero := NewDecFromInt64(0)
+
+	b, err := a.Sub(a)
+	require.NoError(t, err)
+
+	require.True(t, b.IsEqual(zero))
+}
+
+// Property: 1 * a == a
+func testMulLeftIdentity(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	one := NewDecFromInt64(1)
+
+	b, err := one.Mul(a)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(b))
+}
+
+// Property: a * 1 == a
+func testMulRightIdentity(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	one := NewDecFromInt64(1)
+
+	b, err := a.Mul(one)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(b))
+}
+
+// Property: a * b == b * a
+func testMulCommutative(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	b := genDec.Draw(t, "b").(Dec)
+
+	c, err := a.Mul(b)
+	require.NoError(t, err)
+
+	d, err := b.Mul(a)
+	require.NoError(t, err)
+
+	require.True(t, c.IsEqual(d))
+}
+
+// Property: (a * b) * c == a * (b * c)
+func testMulAssociative(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	b := genDec.Draw(t, "b").(Dec)
+	c := genDec.Draw(t, "c").(Dec)
+
+	// (a * b) * c
+	d, err := a.Mul(b)
+	require.NoError(t, err)
+
+	e, err := d.Mul(c)
+	require.NoError(t, err)
+
+	// a * (b * c)
+	f, err := b.Mul(c)
+	require.NoError(t, err)
+
+	g, err := a.Mul(f)
+	require.NoError(t, err)
+
+	require.True(t, e.IsEqual(g))
+}
+
+// Property: (a - b) + b == a
+func testSubAdd(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	b := genDec.Draw(t, "b").(Dec)
+
+	c, err := a.Sub(b)
+	require.NoError(t, err)
+
+	d, err := c.Add(b)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(d))
+}
+
+// Property: (a + b) - b == a
+func testAddSub(t *rapid.T) {
+	a := genDec.Draw(t, "a").(Dec)
+	b := genDec.Draw(t, "b").(Dec)
+
+	c, err := a.Add(b)
+	require.NoError(t, err)
+
+	d, err := c.Sub(b)
+	require.NoError(t, err)
+
+	require.True(t, a.IsEqual(d))
+}
+
+// TODO: Revive this property test with appropriate errors once we need Mul
```

Comment:
I generally don't like to commit commented code, would rather create an issue and paste this code there.

But also won't block this PR either.

Reply:
https://github.com/regen-network/regen-ledger/issues/471